### DISCUSSION TITLE
GLB to Draco compression.

### DISCUSSION THREAD

**Discussion by ovidiu-FX** [donmccurdy/glTF-Transform#347] [General]
Hello my name is Ovidiu, i am FX Artist in Houdini FX, i have a issue with one glb . I export the fbx from Houdini and have size 509 kb. I converted in glb with Cinema4D and when i export it the size increase until 29 MB. The file have animation, and i wish to decrease the size of glb How i proceed to export it with draco? Can you explain to me step by step? Becouse i dont have interactions until now with codes or scripts. Can you help me with that issue? Tanks.

**Comment by donmccurdy** [comment]
Hi @ovidiu-FX!

I've reviewed the models you emailed, but will comment here in case it helps others. In this case the size of the model is ~98% due to morph targets / shape keys. I'm not sure exactly why the GLB uses those for animation — there are other ways to represent this type of animation in glTF/GLB — but perhaps this option is chosen by Cinema4D during the FBX -> GLB conversion, or perhaps it's just how Houdini's animation exporter works? It may be worth investigating in those applications, but I don't know them very well.

As it is, Draco cannot be used to compress morph targets in glTF, it compresses only static geometry, so it won't really help here. Depending on where you want to use this GLB file, other compression options might be supported. You'd need the viewer to support one or both of the glTF extensions below. For example, three.js supports both.

- `KHR_mesh_quantization`: Will reduce size from 29 MB → 20 MB.
- `KHR_mesh_quantization` + `EXT_meshopt_compression`: Will reduce size from 29 MB → 2.5 MB.

I'll email an example created with the second method if you'd like to confirm that it works for what you need. It is worth noting that the compression process does introduce some z-fighting, as the blocks' faces are moving through one another in the same plane... if that's a problem, it may help to introduce some small offsets among the blocks, so their front faces don't coincide.

A quick way to check both Meshopt and Draco compression would be to open the model in https://gltf.report/ and then use the export settings to set up compression:

![Screen Shot 2021-08-26 at 3 18 45 PM](https://user-images.githubusercontent.com/1848368/131043498-9379ebdb-bcd0-42d9-b25e-ee64ff6707e5.png)


With Meshopt you'll get about 4 MB output for this model, but the web viewer doesn't (yet) implement all of the more advanced Meshopt compression options. To get those you can use the [gltfpack](https://github.com/zeux/meshoptimizer/tree/master/gltf#-gltfpack) CLI tool, and this brings it down to about 2.5 MB, which is the best I could achieve on this example.

```
gltfpack -i input.glb -o output.glb -cc
```

I'll email you the example output. If this seems like it will work then I can try to share more step-by-step instructions.

**Reply by ovidiu-FX** [reply]
Hello  Don. Tanks for your reply.  When i try to apply the compressed GLB from that solution, in 8th Wall is dont working. Even in Blender is not recognize it.

Both variants with 4 mb and with 2 mb , dont work, what i must to do?

Blender. 2.89, 2.90, and also on 2.92 say: "Extension EXT_meshopt_compression is not available on this addon version "

And in 8th Wall give me something like this: components:gltf-model:warn THREE.GLTFLoader: setMeshoptDecoder must be called before loading compressed files.

I mention that i wish to work only in 8th Wall . 

Tanks Don.

Ovidiu.

**Reply by donmccurdy** [reply]
It sounds like 8th Wall is using A-Frame VR, perhaps? If that's correct, then support for `EXT_meshopt_compression` was just added to A-Frame two days ago (https://github.com/aframevr/aframe/pull/4910)... it may take a while for 8th Wall to receive that update, so you may need to use less compression for now, or request that they pull in this update sooner.

Unfortunately, without using this type of compression, the best I am able to reduce this model is to about 20 MB using quantization. To do _just_ quantization, bring the model into https://gltf.report/, replace the default script with the one below, run the script, and export without other compression:

```js
import { quantize } from '@gltf-transform/functions';
await document.transform(quantize());
```


Blender and other DCC tools often don't support compressed glTF/GLB files (quantization, above, should usually be OK), so you'll want to reserve the compression for the "final" model. But I'll file an issue (https://github.com/KhronosGroup/glTF-Blender-IO/issues/1460) to see if the addon developers are interested in supporting `EXT_meshopt_compression`.

**Comment by ovidiu-FX** [comment]
Hello Don. yes you have a right. Can you help me please? Because my client
are expecting one result until this night and all day i try,,,and i
try...all possible and impossible tests.....
Can you help me with this ?
I try with Blender but it not work. is increasing the size until 250 MB...
incredible.
Maybe for sure this new update

Tanks.

În vin., 27 aug. 2021 la 20:15, Don McCurdy ***@***.***> a
scris:

**Comment by ovidiu-FX** [comment]
If is ot another way i will try this one. It seems that have 19 MB on this
solution with that code.
 Meanwhile perhaps in future the update will be ok.

Tanks for all Help Don.

Good Day on your side.

Ovidiu.

**Comment by ovidiu-FX** [comment]
Hello Don i talked also with Staff from 8th Wall and this is they answer.:

"We currently do not support EXT_meshopt_compression file types. I have
passed your feedback along to our product team for future consideration.

Another workaround for now would be to export your 3D model without the
meshopt_compression and instead export as a .glb file type. One thing to
keep in mind with 3D models with large file sizes is to make sure you are
using smaller texture sizes. As 4K textures are a bit of an overkill for
mobile devices. I would keep texture files to 512x512 or 1024x1024 to get
an optimized 3D file.

I hope this information is helpful. Please let me know if you are able to
export the file as .glb. "



To clarify on my last email. It looks to be that the request was created 2
days ago https://github.com/aframevr/aframe/pull/4910. 8th Wall uses
A-frame as a 3D framework, so once A-frame releases a new update, 8th wall
will then add this to 8frame (a slightly customized A-frame but for AR).
Then you will be able to use this file type.

The timing here comes down to when A-frame releases an update including the
use of this file type.

Regards,
Ian,

Hi Ovidiu,

You can also take a look at out Draco compression example project as
another possible solution here:
https://www.8thwall.com/playground/draco-compression

Regards,
Ian.

Maybe will implementate soon this update.

Have a Great Weekend.

Ovidiu.

>
